js/src/elements/star-rating.ts

```diff
@@ -0,0 +1,283 @@
+		return ['disabled', 'name', 'value'];
+	}
+
+	constructor() {
+		super();
+		this.attachShadow({ mode: 'open' });
+		this.shadowRoot?.appendChild(template.content.cloneNode(true));
+	}
+
+	get disabled() {
+		return this.hasAttribute('disabled');
+	}
+
+	set disabled(val) {
+		this.setAttribute('disabled', val.toString());
+		const fieldset = this.shadowRoot?.querySelector('fieldset') as HTMLFieldSetElement;
+		fieldset.disabled = val;
+		fieldset.querySelectorAll('input').forEach(input => (input.disabled = val));
+	}
+
+	get name() {
+		return this.getAttribute('name') ?? '';
+	}
+
+	set name(value) {
+		this.setAttribute('name', value);
+	}
+
+	get value() {
+		return Math.max(0, Math.min(parseInt(this.getAttribute('value') || '0'), 5));
+	}
+
+	set value(val) {
+		this.setAttribute('value', val.toString());
+	}
+
+	/** TODO: add form related properties */
+
+	connectedCallback() {
+		this.shadowRoot?.addEventListener('change', this);
+		// arrow function to bind `this` value to star rating in handleFormData
+		this.closest('form')?.addEventListener('formdata', this);
+		this.shadowRoot?.querySelectorAll('input[type="radio"]').forEach(input => {
+			/* Upon initial focus on a brand new star rating, set value to match the currently focused star.
+			Otherwise, the value remains at 0 until user selects the focused star. */
+			input.addEventListener('focus', e => {
+				this.value = parseInt((e.target as HTMLInputElement).value);
+			});
+		});
+	}
+
+	disconnectedCallback() {
+		this.shadowRoot?.removeEventListener('change', this);
+		this.closest('form')?.removeEventListener('formdata', this);
+	}
+
+	attributeChangedCallback(name: string, oldValue: string, newValue: string) {
+		if (oldValue !== newValue) {
+			this.updateContent(name, newValue);
+		}
+	}
+
+	updateContent(name: string, newValue: string) {
+		if (name === 'name') {
+			this.shadowRoot?.querySelectorAll('input[type="radio"]').forEach(input => {
+				input.setAttribute('name', this.name);
+			});
+		}
+		if (name === 'value') {
+			const checkedEl = this.shadowRoot?.querySelector(':checked') as HTMLInputElement;
+			if (checkedEl && this.value === 0) {
+				checkedEl.checked = false;
+			}
+			const toCheck = this.shadowRoot?.querySelector(`[value="${this.value}"]`) as HTMLInputElement;
+			if (toCheck) {
+				toCheck.checked = true;
+				this.updateStarFill(parseInt(newValue));
+			}
+		}
+		if (name === 'disabled') {
+			this.disabled = this.hasAttribute('disabled');
+		}
+	}
+
+	handleEvent(event: Event) {
+		switch (event.type) {
+			case 'change':
+				this.setAttribute('value', (event.target as HTMLInputElement).value);
+				this.dispatchEvent(new Event('change', { bubbles: true }));
+				break;
+			case 'formdata':
+				// https://web.dev/more-capable-form-controls/
+				(event as FormDataEvent).formData.append(this.name, this.value.toString());
+				break;
+		}
+	}
+
+	updateStarFill(newValue: number) {
+		const stars = this.shadowRoot?.querySelectorAll('input');
```

Comment:
think you can use `this.shadowRoot!.query...` here right? To avoid the `if (stars)` below?

Reply:
updated